crates/miden-tx/src/host/link_map.rs

```diff
@@ -297,3 +293,62 @@ enum SetOperation {
     InsertAtHead = 1,
     InsertAfterEntry = 2,
 }
+
+// MEMORY VIEWER
+// ================================================================================================
+
+mod private {
+    pub trait MemoryViewerSeal {}
+}
+
+/// A trait that abstracts over ways to view a process' memory.
+///
+/// More specifically, it allows using a [`LinkMap`] both with a [`ProcessState`], i.e. a process
+/// that is actively executing and also an [`ExecutionOutput`], i.e. a process that has finished
+/// execution.
+///
+/// This should all go away again once we change a LinkMap's implementation to be based on an actual
+/// map type instead of viewing a process' memory directly.
+///
+/// This trait is sealed and should not be implemented by users.
+pub trait MemoryViewer: private::MemoryViewerSeal {
```

Comment:
Why do we need `MemoryViewerSeal`? Is it because we want `MemoryViewer` to be a part of the public interface but don't want anyone else to implement this?

Also, we can't use an `enum` for `MemoryViewer` because of lifetime bounds, right? e.g., something like:

```rust
pub enum MemoryViewer {
    ProcessState(ProcessState),
    ExecutionOutputs(ExecutionOutput),
}
```

Reply:
> Why do we need `MemoryViewerSeal`? Is it because we want `MemoryViewer` to be a part of the public interface but don't want anyone else to implement this?

Yes, that was the rationale. The trait approach was the first thing that came to mind and I didn't consider the enum approach, but it is nicer so I changed to that. Thank you!